caikit/runtime/dump_services.py

```diff
@@ -13,51 +13,69 @@
 # limitations under the License.
 
 # Standard
+from typing import Dict, List, Optional, Union
 import argparse
 import json
 import os
+import shutil
 import sys
 
+# Third Party
+from google.protobuf import descriptor as _descriptor
+from google.protobuf import descriptor_pb2, descriptor_pool
+
 # First Party
+from py_to_proto import descriptor_to_file
+from py_to_proto.utils import safe_add_fd_to_pool
 import alog
 
 # Local
+from ..config.config import get_config
 from ..core.data_model import render_dataobject_protos
-from .service_factory import ServicePackageFactory
-from caikit.config.config import get_config
+from ..core.data_model.dataobject import _AUTO_GEN_PROTO_CLASSES
```

Comment:
big picture q: Where should all this .proto writing logic live?

It looks like we're trying to consolidate it here, so maybe we should kill `data_model.render_dataobject_protos()` and instead have a public `data_model.get_generated_proto_classes()` for this to use.

Ditto on killing `service_package.write_proto_file()` too?

Reply:
Yeah, I was fairly torn about this. The reason I _didn't_ go down this rabbit hole is that fundamentally `data_model` and "dump services" are serving two different purposes that overlap:

* `data_model`: Python structs that can be serialized to protobuf and managed just like any other python objects (NB: dynamic class declaration should be just fine!)
* "dump services": Client-usable protobuf interfaces

The difference comes around how the dynamic nature of the `data_model` classes maps down to `FileDescriptor`s. A `FileDescriptor` is meant to be static (you can easily mutate a `FileDescriptorProto`, but not a `FileDescriptor`), so having the `data_model` classes all live in package-consolidated `FileDescriptor`s would be really challenging since you'd need to know them all apriori.

Because of this gorpiness, I decided to do all of the logic to munch all the existing `Descriptor`/`EnumDescriptor`/`ServiceDescriptor` objects into a consolidated `FileDescriptorProto` -> `FileDescriptor` over here in the `dump_services` module for isolation. This should _not_ be how these protos are created in "normal" operations because it would badly break the extensibility of the data model.

So, this finally comes to the question of whether we kill off `data_model.get_generated_proto_classes` and `service_package.write_proto_file`. I think the former should be kept as it's not strictly about file rendering. I could go either way on the later, but would probably lean towards removing it (although that would be an API breaking change).

Reply:
> This should not be how these protos are created in "normal" operations because it would badly break the extensibility of the data model

Is there a use case for generating .proto files for the data model outside of generating a client for the runtime? I'm just wondering if it's worth keeping the data model's current proto rendering around at all.

> So, this finally comes to the question of whether we kill off `data_model.get_generated_proto_classes` and `service_package.write_proto_file.` I think the former should be kept as it's not strictly about file rendering

I'm confused, I was proposing creating a `data_model.get_generated_proto_classes` method to use instead of accessing a private variable from here. I don't think it exists yet.

> ... although that would be an API breaking change

how far has the ship sailed on caikit ever having a stable api? 😉 
I'd lean towards deleting both of the existing rendering methods as well

Reply:
>I'm confused, I was proposing creating a data_model.get_generated_proto_classes method to use instead of accessing a private variable from here. I don't think it exists yet.

Sorry, bad reading comprehension. You're right that `data_model.render_dataobject_protos()` is basically the same thing as what I have here. I thought about doing all of this as an extra arg in that function, but opted against it because of all the temporary pool muckery.

Reply:
I think the simpler refactor I'm trying to suggest here is just
```
from ..core.data_model.dataobject import _AUTO_GEN_PROTO_CLASSES
```
becomes
```
from  ..core.data_model.dataobject import get_generated_proto_classes
```

But I suppose you could also just rename `_AUTO_GEN_PROTO_CLASSES` -> `AUTO_GEN_PROTO_CLASSES` if we want to access it outside the data model

Reply:
anyway- yeah my care level on this is pretty low. I like the output and consolidating the proto rendering into this file where possible is good. Let's get this merged?

Reply:
Yeah, I agree that importing the private singleton is not great. Will try to fix shortly

Reply:
Done